framework/src/araui/hal/hal_esplcd_impl.c

```diff
@@ -0,0 +1,729 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <math.h>
+#include <semaphore.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sys/ioctl.h>
+#include <tinyara/fs/ioctl.h>
+#include <tinyara/araui_drv.h>
+#include <araui/ui_asset.h>
+#include <araui/ui_commons.h>
+#include "ui_log.h"
+#include "ui_commons_internal.h"
+#include "ui_asset_internal.h"
+#include "ui_core_internal.h"
+#include "hal/ui_hal_interface.h"
+
+#include <iotbus/iotbus_gpio.h>
+
+#define RENDER_TASK_PRIO		125
+//divide to 2 part for flushing, there isn't enough memory for buffer
+#define RENDER_GRAN 			2
+#define UI_IMAGE_BUF_R(image, x, y)   ((image->buf[image->bytes_per_line * y + 3 * x]) >> 3)
+#define UI_IMAGE_BUF_G(image, x, y)   ((((image->buf[image->bytes_per_line * y + 3 * x]) & 0b00000111) << 3) | ((image->buf[image->bytes_per_line * y + 3 * x  + 1] & 0b11100000) >> 5))
+#define UI_IMAGE_BUF_B(image, x, y)   ((image->buf[image->bytes_per_line * y + 3 * x + 1]) & 0b00011111)
+#define UI_IMAGE_BUF_A(image, x, y)   (image->buf[image->bytes_per_line * y + 3 * x + 2])
+#define UI_INTERPOLATION(color1, color2, weight)  (color1 * weight + color2 * (1 - weight))
+
+#define DEBUG_RENDER_TIME       0
+#define DEBUG_BUTTON_SIM		0
+
+uint32_t _ui_hal_get_buf_color(ui_image_asset_body_t *body, int32_t x, int32_t y);
+uint32_t _ui_hal_get_buf_interpolated_color(ui_image_asset_body_t *body, double x, double y);
+
+typedef struct {
+	int x1;
+	int y1;
+	int x2;
+	int y2;
+} area_t;
+
+static uint16_t *g_framebuf;
+static uint16_t *g_renderbuf;
+static const int32_t FRAMEBUF_SIZE = (CONFIG_UI_DISPLAY_WIDTH * CONFIG_UI_DISPLAY_HEIGHT) * 2;
+static pid_t g_tpid;
+static sem_t g_mutex;
+static sem_t g_notify;
+static area_t g_rendering_area;
+static ui_rect_t g_viewport = {0, };
+static int g_ui_fd;
+
+#if DEBUG_BUTTON_SIM
+static iotbus_gpio_context_h g_button;
+static pid_t g_timerpid;
+static struct timespec g_prev_click;
+static struct timespec g_prev_rel;
+static int g_pressed;
+static int g_longpressing;
+static int g_checkclick;
+
+static int longpress_check_task(int argc, char *argv[])
+{
+	uint32_t dt;
+	struct timespec now;
+	memset(&now, 0, sizeof(struct timespec));
+
+	while(1) {
+
+		if (g_pressed) {
+			clock_gettime(CLOCK_MONOTONIC, &now);
+			dt = ((now.tv_sec - g_prev_click.tv_sec) * 1000) + ((now.tv_nsec - g_prev_click.tv_nsec) / 1000000);
+			if (dt > 400) {
+				// Send Long press event here!
+				printf("Long pressing!\n");
```

Comment:
It'd better to use lldbg or its variants as per its purpose.

Reply:
You're right! I'll remove this printf